lib/src/impl_ffi/impl_ffi.utils.dart

```diff
@@ -14,27 +14,24 @@
 
 part of impl_ffi;
 
-/// Function that can't be inlined to be used for preventing [obj] from being
-/// garbage collected.
-@pragma('vm:never-inline')
-Object _finalizerReachabilityFence(Object obj) {
-  return obj;
-}
+/// Wrapper around [EVP_PKEY] which attaches finalizer and ensure that the
+/// [ffi.Finalizable] is kept in scope while the [EVP_PKEY] is used.
+class _EvpPKey implements ffi.Finalizable {
+  /// We don't really have an estimate of how much space the EVP_PKEY structure
+  /// takes up, but if we make it some non-trivial size then hopefully the GC
+  /// will prioritize freeing them.
+  static const _externalSizeEstimate = 4096;
+
+  static final _finalizer =
+      ffi.NativeFinalizer(ssl.addresses.EVP_PKEY_free.cast());
 
-/// Wrapper around [EVP_PKEY] which attaches finalizer and ensures that the
-/// [_finalizerReachabilityFence] is used after usage.
-///
-/// The [_finalizerReachabilityFence] ensures that the [_EvpPKey] is not garbage
-/// collected while the wrapped key is use. Thus, we can be sure the finalizer
-/// won't be called while the wrapped key is in use.
-class _EvpPKey {
   final ffi.Pointer<EVP_PKEY> _pkey;
 
   /// Allocate new [EVP_PKEY], attach finalizer and return the wrapped key.
   factory _EvpPKey() {
-    final _pkey = ssl.EVP_PKEY_new();
-    _checkOp(_pkey.address != 0, fallback: 'allocation failure');
-    return _EvpPKey.wrap(_pkey);
+    final pkey = ssl.EVP_PKEY_new();
+    _checkOp(pkey.address != 0, fallback: 'allocation failure');
+    return _EvpPKey.wrap(pkey);
   }
 
   /// Wrap existing [EVP_PKEY], this will attach a finalizer.
```

Comment:
Do you ever want to early-free a key?

Usually native resources have a `_released`/`_deleted` internal boolean that is being check before every method and a `release`/`delete` method that removes the finalizer and calls the native "free" early.
(Even better if that works also with Arena's: https://github.com/dart-lang/sdk/issues/49906 )

Reply:
Not really, I don't think I have any use-cases where I release early.